File: solutions/javascript/matching-brackets/1/matching-brackets.js
//
// This is only a SKELETON file for the 'Matching Brackets' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const isPaired = (brackets) => {
  const stack = [];
  const bracketsMap = new Map()
  bracketsMap.set("[", "]");
  bracketsMap.set("{", "}");
  bracketsMap.set("(", ")");
  const openBrackets = "[{(";
  const closeBrackets = "]})";
  for (const element of brackets) {
    if (openBrackets.includes(element)) {
      stack.push(element)
    }
    if (closeBrackets.includes(element)) {
      if (!stack.length) return false;
      const openBracket = stack.pop();
      if (bracketsMap.get(openBracket) !== element) return false;
    }
  }
  return !stack.length
};

File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    open_brackets = "[{("
    close_brackets = "]})"
    now_open_brackets = []
    to_close_brackets = []
    for input in input_string:
        if input in open_brackets:
            open_bracket_index = open_brackets.index(input)
            now_open_brackets.append(input)
            to_close_brackets.append(close_brackets[open_bracket_index])
        elif input in close_brackets:
            if len(to_close_brackets) == 0:
                return False
            to_close_bracket = to_close_brackets.pop()
            if input != to_close_bracket:
                return False
            now_open_brackets.pop() 
    return len(now_open_brackets) == 0 and len(to_close_brackets) == 0
        
